### PR TITLE
Introduce the generics pre-processing pass for concepts and close many static[T] issues

### PR DESCRIPTION
**Pull request by zah**

### CODE REVIEW

**Comment by Araq** [comment]
All in all, looks great. Do the concept heavy projects out there work with it?

**Comment by zah** [comment]
I've tested only `linalg`, which is not affected. Out test suite has grown considerably during the previous regression fixing effort.

File: compiler/sigmatch.nim [hunk 1]
```diff
@@ -302,8 +302,30 @@ proc describeArgs*(c: PContext, n: PNode, startIdx = 1;
     add(result, argTypeToString(arg, prefer))
     if i != sonsLen(n) - 1: add(result, ", ")
 
-proc typeRel*(c: var TCandidate, f, aOrig: PType,
-              flags: TTypeRelFlags = {}): TTypeRelation
+proc typeRelImpl*(c: var TCandidate, f, aOrig: PType,
+                  flags: TTypeRelFlags = {}): TTypeRelation
+
+var nextTypeRel = 0
```
Comment (Araq) [inline compiler/sigmatch.nim]: This should be in a 'when traceTypeRel' too, right?

File: compiler/semdata.nim [hunk 2]
```diff
@@ -40,6 +40,11 @@ type
     bracketExpr*: PNode       # current bracket expression (for ^ support)
     mapping*: TIdTable
 
+  TMatchedConcept* = object
```
Comment (dom96) [inline compiler/semdata.nim]: The 'T' prefix is deprecated now.

File: compiler/semcall.nim [hunk 3]
```diff
@@ -446,7 +446,8 @@ proc explicitGenericSym(c: PContext, n: PNode, s: PSym): PNode =
 proc explicitGenericInstantiation(c: PContext, n: PNode, s: PSym): PNode =
   assert n.kind == nkBracketExpr
   for i in 1..sonsLen(n)-1:
-    n.sons[i].typ = semTypeNode(c, n.sons[i], nil)
+    let e = semExpr(c, n.sons[i])
```
Comment (Araq) [inline compiler/semcall.nim]: IIRC 'semTypeNode' can have different lookup rules than 'semExpr', types are preferred in a type context, perhaps?

File: compiler/semdata.nim [hunk 4]
```diff
@@ -40,6 +40,11 @@ type
...
+    candidateType*: PType
+    prev*: ptr TMatchedConcept
```
Comment (dom96) [inline compiler/semdata.nim]: Pointers are unsafe, why are you using them instead of refs?

File: compiler/semdata.nim (hunk 2 above)
Comment (Araq) [inline compiler/semdata.nim]: Doesn't matter in the compiler though, the compiler never adhered to NEP1.

File: compiler/semdata.nim (hunk 4 above)
Comment (Araq) [inline compiler/semdata.nim]: Speed?

File: compiler/sigmatch.nim (hunk 1 above)
Comment (zah) [inline compiler/sigmatch.nim]: It could be, but I didn't bother to try to avoid a single extra global variable.

File: compiler/semcall.nim (hunk 3 above)
Comment (zah) [inline compiler/semcall.nim]: Well, `semTypeNode` doesn't cut it here anymore. These are the explicit generic parameters of a proc call, so they can be static values as well.

Comment (Araq) [inline compiler/semcall.nim]: semExprWithType then?

**Comment by zah** [comment]
Rebased against master in order to fix some conflicts.

**Pull request merged**